specs/WebResourceRequested-CustomScheme.md

```diff
@@ -0,0 +1,295 @@
+``` c#
+CoreWebView2EnvironmentOptions options = new CoreWebView2EnvironmentOptions();
+string[] allowedOriginsList = new string[1];
+allowedOriginsList[0] = "https://*.example.com";
+
+options.CustomSchemeRegistrations.Add(new CoreWebView2CustomSchemeRegistration("custom-scheme", true /* isSecure */, allowedOriginsList));
+options.CustomSchemeRegistrations.Add(new CoreWebView2CustomSchemeRegistration("custom-scheme-not-in-allowed-origins", true /* isSecure */, null));
+...
+string customScheme = "custom-scheme";
+webView.CoreWebView2.AddWebResourceRequestedFilter(
+        customScheme + ":*", All);
+webView.CoreWebView2.WebResourceRequested += delegate (
+            object sender, CoreWebView2WebResourceRequestedEventArgs e) {
+        CoreWebView2WebResourceResponse response;
+        CoreWebView2WebResourceRequest request = e.Request;
+        String uri = request.uri;
+        if (uri.StartsWith(customScheme + "://"))
+        {
+            string assetsFilePath = L"data/";
+            assetsFilePath += uri.Substring(customScheme.Length + 3);
+            FileStream fileStream = new FileStream(assetsFilePath, FileMode.Open, FileAccess.Read);
+
+            if (stream)
+            {
+              e.Response = webView.CoreWebView2.CreateWebResourceResponse(fileStream, 200, "OK", L"Content-Type: application/json\nAccess-Control-Allow-Origin: *");
+            }
+            else
+            {
+                e.Response = webView.CoreWebView2.CreateWebResourceResponse(null, 404, "Not Found", L"");
+            }
+        }
+}
+
+webView.CoreWebView2.Navigate(L"https://www.example.com");
+
+// The following XHR will execute in the context of https://www.example.com page and will succeed.
+// WebResourceRequested event will be raised for this request as *.example.com is in the allowed origin list of custom-scheme.
+// Since the response header provided in WebResourceRequested handler allows all origins for CORS
+// the XHR succeeds.
+webView.CoreWebView2.ExecuteScriptAsync("var oReq = new XMLHttpRequest();\
+                                        oReq.addEventListener(\"load\", reqListener);\
+                                        oReq.open(\"GET\", \"custom-scheme://example-data.json\");\
+                                        oReq.send();");
+// The following XHR will fail because *.example.com is in the not allowed origin list of custom-scheme2.
+// The WebResourceRequested event will not be raised for this request.
+webView.CoreWebView2.ExecuteScriptAsync("var oReq = new XMLHttpRequest();\
+                                        oReq.addEventListener(\"load\", reqListener);\
+                                        oReq.open(\"GET\", \"custom-scheme-not-in-allowed-origins://example-data.json\");\
+                                        oReq.send();");
+```
+
+``` cpp
```

Comment:
Note: C++/WinRT is preferred over WRL. The code will look much closer to C# and thus be easier to read.

Reply:
Result of our sample app requiring WRL::Callback which C++/winrt and WIL do not have a replacement for.

Reply:
Is this something customers have to deal with? Or is it just an "us" problem?

Reply:
TLDR its a problem for a subset of our customers. WebView2 has a .NET API (most popular by far), WinRT API, and a COM API (the one in this sample). If a WebView2 end dev is using our COM API and using our events or async methods then they need to implement our async method completion delegate interfaces and our event handler delegate interfaces. The WRL Callback feature makes that very easy. Or they can implement the interface manually. C++/winrt and WIL don't provide a substitute for WRL Callback.